### PR TITLE
Listing rules service created

### PR DESCRIPTION
**Pull request by abdura52**
ListingRules service created

### CODE REVIEW

File: src/Training.AirBnb.Clone.Backend/Backend Project/Domain/Entities/ListingRules.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    public bool SmokingAllowed { get; set; }
+
+    public bool CommercialFilmingAllowed { get; set; }
+
+    public TimeOnly? CheckInTimeStart { get; set; }
+
+    public TimeOnly? CheckInTimeEnd { get; set; }
+
+    public TimeOnly CheckOutTime { get; set; }
+
```
Comment (Ilxom-22) [inline src/Training.AirBnb.Clone.Backend/Backend Project/Domain/Entities/ListingRules.cs:22]: CheckOutTime nullable bo'mi qobti.

**Review by Ilxom-22 — CHANGES_REQUESTED** [review]

File: src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        if (listingRules.CheckInTimeStart is null && listingRules.CheckInTimeEnd is not null)
+        {
+            throw new EntityValidationException<ListingRules>("Check-in end time must be left unspecified when the check-in start time is missing or null.");
+        }
+
+        if (listingRules.CheckInTimeStart is not null
+            && (listingRules.CheckInTimeEnd is null
+            || (listingRules.CheckInTimeEnd - listingRules.CheckInTimeStart) >= TimeSpan.FromHours(2)))
+        {
```
[the remark this answers is not in the document]
Comment (abdura52) [reply src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs:106]: bitta character nimalarga olib kemidi 🤣

File: src/Training.AirBnb.Clone.Backend/Backend Project/Domain/Entities/ListingRules.cs (hunk 1 above)
Comment (abdura52) [reply src/Training.AirBnb.Clone.Backend/Backend Project/Domain/Entities/ListingRules.cs:22]: CheckOutTime nullable bo'lmagan saytda ham

File: src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+{
+    private readonly IDataContext _context;
+
+    public ListingRulesService(IDataContext context)
+    {
+        _context = context;
+    }
+
+    public async ValueTask<ListingRules> CreateAsync(ListingRules entity, bool saveChanges = true, CancellationToken cancellationToken = default)
+    {
```
Comment (Abdurahmon0412) [inline src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs:19]: entityni listingrulesga o'zgartirsayiz bo'ladi

File: src/Training.AirBnb.Clone.Backend/Backend Project/Domain/Entities/ListingRules.cs (hunk 1 above)
Comment (Ilxom-22) [reply src/Training.AirBnb.Clone.Backend/Backend Project/Domain/Entities/ListingRules.cs:22]: Ha CheckInEndTime bilan adashtirvorimman.

File: src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+        if (saveChanges)
+            await _context.SaveChangesAsync();
+
+        return entity;
+
+    }
+
+    public async ValueTask<ListingRules> DeleteAsync(Guid id, bool saveChanges = true, CancellationToken cancellationToken = default)
+    {
```
Comment (Abdurahmon0412) [inline src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs:32]: Delete methodi eng oxirida updatedan keyin bo'adi

File: src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        Validate(entity);
+
+        await _context.ListingRules.AddAsync(entity, cancellationToken);
+
+        if (saveChanges)
+            await _context.SaveChangesAsync();
+
+        return entity;
+
```
Comment (Abdurahmon0412) [inline src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs:28]: returindan keyin qator tashlamasayizham bo'ladi redibility uchun

File: src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    public IQueryable<ListingRules> Get(Expression<Func<ListingRules, bool>> predicate)
+        => _context.ListingRules.Where(predicate.Compile()).AsQueryable();
+
+    public ValueTask<ICollection<ListingRules>> GetAsync(IEnumerable<Guid> ids, CancellationToken cancellationToken = default)
+        => new ValueTask<ICollection<ListingRules>>(_context.ListingRules.Where(listingRules => ids.Contains(listingRules.Id)).ToList());
+
+    public async ValueTask<ListingRules> GetByIdAsync(Guid id, CancellationToken cancellationToken = default)
+        => await _context.ListingRules.FindAsync(id, cancellationToken)
+            ?? throw new EntityNotFoundException<ListingRules>("Entity not found with this id");
```
Comment (Abdurahmon0412) [inline src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs:55]: hamma joyda _contextni o'rniga bitta undeleted methodni yozib o'shaniishlatsayiz bo'ladi

File: src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+    public async ValueTask<ListingRules> UpdateAsync(ListingRules entity, bool saveChanges = true, CancellationToken cancellationToken = default)
+    {
```
Comment (Abdurahmon0412) [inline src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs:58]: bu yerdaham entitylarni almashtirib yuborasiz

File: src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+    private void Validate(ListingRules listingRules)
+    {
+        var minGuestsCount = 1;
+
+        if (listingRules.GuestsCount < minGuestsCount)
+        {
+            throw new EntityValidationException<ListingRules>("Guests count isn't valid!");
+        }
+
```
Comment (Abdurahmon0412) [inline src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingRulesService.cs:97]: ifda o'zi hamma narsa bir qatorda ekanku scop nega kerak 
bu xato emas prosta redibility uchun

**Review by Abdurahmon0412 — CHANGES_REQUESTED** [review]

**Review by Habiba67y — APPROVED** [review]

**Review by Ilxom-22 — APPROVED** [review]

**Review by Abdurahmon0412 — APPROVED** [review]

**Pull request merged**